src/fmu/tools/qcforward/_blockedwells_vs_gridprops.py

```diff
@@ -0,0 +1,328 @@
+    ):
+        """Main routine for evaluating blockedwells vs gridproperties
+
+        The routine depends on existing XTGeo functions for this purpose.
+
+        Args:
+            data (dict or str): The input data either as a Python dictionary or
+                a path to a YAML file
+            reuse (bool or list): Reusing some "timeconsuming to read" data in the
+                instance. If True, then grid and gridprops will be reused as default.
+                Alternatively it can be a list for more fine grained control, e.g.
+                ["grid", "gridprops", "bwells"]
+            project (Union[object, str]): For usage inside RMS, None if running files
+
+        """
+        self._data = self.handle_data(data, project)
+        self._validate_input(self._data, project)
+
+        QCC.verbosity = self._data.get("verbosity", 0)
+
+        # parse data that are special for this check
+        QCC.print_info("Parsing additional data...")
+        self.ldata = _LocalData()
+        self.ldata.parse_data(self._data)
+
+        # now need to retrieve blocked properties and grid properties from the "compare"
+        # dictionary:
+        wsettings = {"lognames": list(self.ldata.compare.keys())}
+
+        if project:
+            # inside RMS, get gridprops implicitly from compare values
+            self._data["gridprops"] = list(self.ldata.compare.values())
+
+        if not isinstance(self.gdata, QCData):
+            self.gdata = QCData()
+
+        self.gdata.parse(
+            data=self._data, reuse=reuse, project=project, wells_settings=wsettings
+        )
+
+        dfr, comb = self.compare_bw_props()
+
+        QCC.print_debug(f"Results: \n{dfr}")
+        status = self.evaluate_qcreport(
+            dfr, "blocked wells vs grid props", stopaction=False
+        )
+
+        # make it possible to print the full dataframe, either some (.e.g the failing)
+        # or all
+        if self.ldata.show_data:
+            if "fail" in self.ldata.show_data and status == "STOP":
+                dfr_stop = dfr[dfr["STATUS"] == "STOP"]
+                wells = list(dfr_stop["WELL"])
+                if "all" in wells:
+                    wells.remove("all")
+                print(f"Failing wells are these {wells}:")
+                print(comb[comb["WELLNAME"].isin(wells)].to_string())
+            elif "warn" in self.ldata.show_data:
+                dfr_warn_or_stop = dfr[dfr["STATUS"] != "OK"]
+                wells = list(dfr_warn_or_stop["WELL"])
+                if "all" in wells:
+                    wells.remove("all")
+                if wells:
+                    print(f"Wells with WARN or STOP are these {wells}:")
+                    print(comb[comb["WELLNAME"].isin(wells)].to_string())
+                else:
+                    print("No wells reported WARN or STOP")
+
+            else:
+                print(f"Wells with WARN or STOP or OK are:")
+                print(comb.to_string())
```

Comment:
This could probably be shortened a bit. Something like:

```
        if self.ldata.show_data:
            if self.ldata.show_data in ["fail", "warn"]:
		mask = (dfr["STATUS"] == "STOP") if self.ldata.show_data == "fail" else (dfr["STATUS"] != "OK")
                wells = [well for well in dfr[mask]["WELL"].unique() if well != "all"]
		if wells:
                    print(f"Wells within {self.ldata.show_data} criteria are these {wells}:")
                    print(comb[comb["WELLNAME"].isin(wells)].to_string())
                else:
                    print("No wells within {self.ldata.show_data} criteria")
            else:
                print(f"Wells with WARN or STOP or OK are:")
                print(comb.to_string())
```

Reply:
After testing on Gullfaks with many wells, I realize that if the user select `fail `or `warn `as the desired output, we should filter away rows that is flagged as "MATCH" to highlight the inconsistent data!

Reply:
> Overall I like the functionality, it is quite fast to use and offers a quick qc to secure that your parameters have been conditioned to 👍 🙂
> 
> For facies this can be a good tool if the user have created a facies parameter by merging different outputs from different facies modelling jobs, this can easily lead to mismatch between the BW log and the grid parameter.
> 
> I think it might be useful to include some filter options though. e.g, when looking at Sw match it would be beneficial to filter away the water zone. Also for other parameters it would be nice being able to ignore parts of the BW logs if it is not of interest.

Making such filters can be a challenge if the field have many contact compartments, and/or individual wells are influenced by production. One possibly simpler option is to create a single filter value for a group of wells (but I need in case look into the details). A side note is that SW is usually not conditioned to blocked wells; the main purpose with this is blocked wells that are used in hard-conditioning (porosity, permeability, facies, ...) 

> 
> Mainly the code looks ok, I have made some comments. My main concern from a user perspective is the actions. At the moment there is not enough checks on the format of these and it is easy to make mistakes. For this qcforward test maybe we should change the format of the actions, the user will e.g. never want to look at a match above a certain percentage so we actually don't need the "<" sign. Maybe actions could be a dictionary then it is easier to check the input and we avoid having so many checks on a string? What do you think?

In some other qcforward jobs both > and < are applied. So perhaps consistency between jobs is good? I also think the signs add readability. Otherwise I agree that better tests on the actions is beneficial, so I will take a look

> 
> ```
> actions = {
>     "warn": {"any": 90, "all": 95}, 
>     "stop": {"any": 80, "all": 90},
> }
> ```
> 
> As a last comment it could be nice to add typehints, probably we would like to include mypy into the CI for fmu.tools at some point 🙂

Yup, adding type hints is a good idea (I already have this in xtgeo for all new code)